sdk/dotnet/Pulumi/Deployment/Deployment_Complete.cs

```diff
@@ -0,0 +1,93 @@
+            Resource resource, Func<Task<(string urn, string id, Struct data)>> action)
+        {
+            // IMPORTANT!  This function is Task-returning, but must not actually be `async` itself.
+            // We have to make sure we run this synchronously directly when the constructor runs
+            // since this will set all our output fields.  We need those fields assigned by the time
+            // the constructor returns.
+
+            return CompleteResourceAsync(
+                resource, action, OutputCompletionSource.GetSources(resource));
+        }
+
+        private async Task CompleteResourceAsync(
+            Resource resource, Func<Task<(string urn, string id, Struct data)>> action,
+            ImmutableDictionary<string, IOutputCompletionSource> completionSources)
+        {
+            // Run in a try/catch/finally so that we always resolve all the outputs of the resource
+            // regardless of whether we encounter an errors computing the action.
+            try
+            {
+                var response = await action().ConfigureAwait(false);
+                completionSources[Constants.UrnPropertyName].SetStringValue(response.urn, isKnown: true);
+                if (resource is CustomResource customResource)
+                {
+                    var id = response.id ?? "";
+                    completionSources[Constants.IdPropertyName].SetStringValue(id, isKnown: id != "");
+                }
+
+                // Go through all our output fields and lookup a corresponding value in the response
+                // object.  Allow the output field to deserialize the response.
+                foreach (var (fieldName, completionSource) in completionSources)
+                {
+                    if (fieldName == Constants.UrnPropertyName || fieldName == Constants.IdPropertyName)
+                    {
+                        // Already handled specially above.
+                        continue;
+                    }
+
+                    // We process and deserialize each field (instead of bulk processing
+                    // 'response.data' so that each field can have independent isKnown/isSecret
+                    // values.  We do not want to bubble up isKnown/isSecret from one field to the
+                    // rest.
+                    if (response.data.Fields.TryGetValue(fieldName, out var value))
+                    {
+                        var converted = Converter.ConvertValue(
+                            $"{resource.GetType().FullName}.{fieldName}", value, completionSource.TargetType);
+                        completionSource.SetValue(converted);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                // Mark any unresolved output properties with this exception.  That way we don't
+                // leave any outstanding tasks sitting around which might cause hangs.
```

Comment:
in general, no.  Basically, since we have Deployment.RunAsync which roots our tasks and awaits them, we will only see and report the first exception it sees.